src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,453 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseRecursivePatterns
+{
+    public partial class UseRecursivePatternsTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+           => new CSharpUseRecursivePatternsCodeRefactoringProvider();
+
+        [Fact]
+        public async Task Test1()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.P1 == 1 [||]&& b.P2 == 2;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is { P1: 1, P2: 2 };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test2()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.MetadataName == string.Empty &&
+            b.ContainingType == null &&
+            b.ContainingNamespace != null &&
+            b.ContainingNamespace.Name == string.Empty [||]&&
+            b.ContainingNamespace.ContainingNamespace == null;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is
+        {
+            MetadataName: string.Empty,
```

Comment:
`string.Empty` is not a valid pattern, right?

Reply:
it's not. I skipped constant checking for testing, should be enabled eventually. I'll adjust the tests afterwards.